File: codes/java/chap09/answer_A70.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		// 입력
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int M = sc.nextInt();
		int[] A = new int[N + 1];
		int[] X = new int[M + 1];
		int[] Y = new int[M + 1];
		int[] Z = new int[M + 1];
		for (int i = 1; i <= N; i++) {
			A[i] = sc.nextInt();
		}
		for (int i = 1; i <= M; i++) {
			X[i] = sc.nextInt();
			Y[i] = sc.nextInt();
			Z[i] = sc.nextInt();
		}

		// 인접 리스트를 생성하고 그래프에 에지를 추가
		ArrayList<Integer>[] G = new ArrayList[1 << N];
		for (int i = 0; i < (1 << N); i++) {
			G[i] = new ArrayList<Integer>();
		}
		for (int i = 0; i < (1 << N); i++) {
			for (int j = 1; j <= M; j++) {
				int nextState = getNext(N, i, X[j], Y[j], Z[j]);
				G[i].add(nextState);
			}
		}

		// 시작과 끝의 정점 번호 찾기
		int start = 0;
		for (int i = 1; i <= N; i++) {
			if (A[i] == 1) {
				start += (1 << (i - 1));
			}
		}
		int goal = (1 << N) - 1;

		// 배열 초기화／시작 시점을 큐에 넣는다
		int[] dist = new int[1 << N];
		Arrays.fill(dist, -1);
		dist[start] = 0;
		Queue<Integer> Q = new LinkedList<>();
		Q.add(start);

		// 너비 우선 탐색
		while (Q.size() >= 1) {
			int pos = Q.remove();
			for (int i = 0; i < G[pos].size(); i++) {
				int nex = G[pos].get(i);
				if (dist[nex] == -1) {
					dist[nex] = dist[pos] + 1;
					Q.add(nex);
				}
			}
		}

		// 답을 출력
		System.out.println(dist[goal]);
	}
	
	// 정점 pos의 상태에서 '램프 x, y, z의 상태'를 반전시켰을 때 정점 번호를 반환하는 함수
	// (책에 있는 코드에서는 'vertex pos의 상태에서 idx번째 작업을 수행했을 때'의 정점 번호를 반환하는 함수를 구현하고 있지만, 책에  있는 코드에서는 이와는 조금 다르게 구현하고 있습니다.)
	static int getNext(int N, int pos, int x, int y, int z) {
		// pos의 2진법 표기법을 사용하여 정점 pos가 나타내는 램프의 상태 state를 계산합니다.
		// (2진법으로 변환하는 방법은 1.4절 참조)
		int[] state = new int[N + 1];
		for (int i = 1; i <= N; i++) {
			int wari = (1 << (i - 1));
			state[i] = (pos / wari) % 2;
		}

		// 램프 x, y, z 상태 반전
		state[x] = 1 - state[x];
		state[y] = 1 - state[y];
		state[z] = 1 - state[z];

		// 10진법으로 변환하는 방법도 1.4절을 참조
		int ret = 0;
		for (int i = 1; i <= N; i++) {
			if (state[i] == 1) {
				ret += (1 << (i - 1));
			}
		}
		return ret;
	}
}

// 참고 1: 이 문제에 대한 더 간결한 구현도 있으니, 원한다면 answer_A70_extra.py도 참고하시기 바랍니다.

File: codes/cpp/chap09/answer_A70.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int N, A[19];
int M, X[109], Y[109], Z[109];
int dist[1033];
vector<int> G[1033];

// 노드 pos의 상태에서 'idx 번째 조작'을 수행했을 때의 노드 번호를 반환한다
int GetNext(int pos, int idx) {
	int State[19]; // State[i]는 램프 i의 상태

	// 2진법으로 고치는 방법은 1.4절을 참조
	for (int i = 1; i <= N; i++) {
		int wari = (1 << (i - 1));
		State[i] = (pos / wari) % 2;
	}
	State[X[idx]] = 1 - State[X[idx]]; // 램프 X[idx]를 반전
	State[Y[idx]] = 1 - State[Y[idx]]; // 램프 Y[idx]를 반전
	State[Z[idx]] = 1 - State[Z[idx]]; // 램프 Z[idx]를 반전

	// 10진법으로 변환하는 방법도 1.4절을 참조
	int ret = 0;
	for (int i = 1; i <= N; i++) {
		if (State[i] == 1) ret += (1 << (i - 1));
	}
	return ret;
}

int main() {
	// 입력
	cin >> N >> M;
	for (int i = 1; i <= N; i++) cin >> A[i];
	for (int j = 1; j <= M; j++) cin >> X[j] >> Y[j] >> Z[j];

	// 그래프에 에지를 추가
	for (int i = 0; i < (1 << N); i++) {
		for (int j = 1; j <= M; j++) {
			int NextState = GetNext(i, j);
			G[i].push_back(NextState);
		}
	}

	// 시작 지점, 목표 지점의 노드 번호를 구한다
	int Goal = (1 << N) - 1;  // (1<<N)은 '2의 N 제곱'
	int Start = 0;
	for (int i = 1; i <= N; i++) {
		if (A[i] == 1) Start += (1 << (i - 1));
	}

	// 배열 초기화／시작 시점을 큐에 넣는다
	queue<int> Q;
	for (int i = 0; i < (1 << N); i++) dist[i] = -1;
	dist[Start] = 0;
	Q.push(Start);

	// 너비 우선 탐색
	while (!Q.empty()) {
		int pos = Q.front(); Q.pop();
		for (int i = 0; i < G[pos].size(); i++) {
			int nex = G[pos][i];
			if (dist[nex] == -1) {
				dist[nex] = dist[pos] + 1;
				Q.push(nex);
			}
		}
	}

	// 답을 출력
	cout << dist[Goal] << endl;
	return 0;
}
